### PR TITLE
Riemannian Gradient Descent on the circle

### PR DESCRIPTION
**Pull request by bollu**
I shall be porting a [notebook of mine](https://github.com/bollu/bollu.github.io/blob/master/N_Sphere_SGD.ipynb) that performs riemannian gradient descent on the sphere to `SciLean`.

### CODE REVIEW

File: examples/CircleOptimisation.lean [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+-- sees that it is manifestly smooth?
+-- I took a look at `Pendulum2D.lean` which seems to not compile anymore.
+def A (x : ℝ^{2}) : ℝ^{2} := 
+  ⊞ i, 
+    let x₀ := x.get 0
+    let x₁ := x.get 1
+    let x' := #[x₀ + x₁, x₀ - x₁];
+    have H : x'.size = 2 := by simp;
+    let x' := x'.get (H ▸ i.toFin)
+    x'
```
Comment (bollu) [inline examples/CircleOptimisation.lean:27]: I already run into trouble, since I am not sure how to build the matrix

```
[1  1]
[-1 1]
``` 

and multiply it with the input vector `x` such that `SciLean` can automatically see that it is smooth. It is clear that perform index manipulation is not the way to go, but I am unsure what the correct API is.

Comment (lecopivo) [reply examples/CircleOptimisation.lean:27]: You can define `A` simply as
```
def A (x : ℝ^{2}) : ℝ^{2} := 
  ⊞ i, if i = 0 then x[0] + x[1] else x[0] - x[1]
```

File: examples/CircleOptimisation.lean [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+/--
+failed to synthesize instance
+  IsSmooth fun x => A x
+-/
+function_properties A (x : ℝ^{2})
+argument x
+  IsLin := sorry_proof,
+  HasAdjoint := sorry_proof,
+  IsSmooth,
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:36]: With the above definition of `A` all proofs are done automatically, so you can just write `IsLin, HasAdjoint, IsSmooth,`

File: examples/CircleOptimisation.lean [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  abbrev ∂ := λ dx => A dx by sorry_proof,
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:37]: You can either specify the differential manually
```
  abbrev ∂ := λ dx => A dx by unfold A; fun_trans,
```
or generate automatically
```
  def ∂ by unfold A; fun_trans; simp,
```
When using `def`, AD will rewrite `∂ x, A x` to `A.arg_x.differential`.
When using `abbrev`, AD will rewrite `∂ x, A x` to `fun x dx => A dx` or `fun x dx => ⊞ i, if i = 0 then dx[0] + dx[1] else dx[0] - dx[1]` depending if you used option 1 or 2.

Effectively `def` or `abbrev` affects what appears on the rand hand side of the theorem `A.arg_x.differential_simp` and `A.arg_x.differential_simp'`

File: examples/CircleOptimisation.lean [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  noncomputable abbrev † := λ x' => A† x' by sorry_proof,
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:38]: This is currently broken as it chokes on computing `(fun x => x[0])†

btw. you can observe the progress of AD with
```
set_option trace.Meta.Tactic.fun_trans.rewrite true in
```

File: examples/CircleOptimisation.lean [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  HasAdjDiff,
+  noncomputable abbrev ∂† := λ dx' => A† dx' by sorry_proof
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:40]: This can be 
```
  abbrev ∂† by unfold A; fun_trans; simp
```

Unfortunately it generates this monstrosity
```
fun dx' =>
    sum fun i =>
      setElem 0 0 (if i = 0 then (dx'[i], 0) else (0, dx'[i])).fst +
          setElem 0 1 (if i = 0 then (dx'[i], 0) else (0, dx'[i])).fst +
        (setElem 0 0 (if i = 0 then (dx'[i], 0) else (0, dx'[i])).snd +
          setElem 0 1 (-(if i = 0 then (dx'[i], 0) else (0, dx'[i])).snd))
```
Computing just the transpose would have the same problem. In general, transposing functions with element access is really difficult and I'm still struggling how to approach it.

**Comment by lecopivo** [comment]
The definition of `A` and its properties should looks like this:
```
def A (x : ℝ^{2}) : ℝ^{2} := 
  ⊞ i, if i = 0 then x[0] + x[1] else - x[0] + x[1]

function_properties A (x : ℝ^{2})
argument x
  IsLin, HasAdjoint, IsSmooth, HasAdjDiff,
  abbrev ∂ := λ dx => A dx by unfold A; fun_trans,
  def † := λ x' => ⊞ i, if i = 0 then x'[0] - x'[1] else x'[0] + x'[1] by sorry,
  abbrev ∂† by unfold adjointDifferential; fun_trans; fun_trans; simp
```

Once the adjoint is fixed then you probably write ` def † by unfold A; fun_trans`.

File: examples/CircleOptimisation.lean [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+function_properties loss (x : ℝ^{2})
+argument x
+  IsSmooth, HasAdjDiff,
+  abbrev ∂ := λ dx => (2 : ℝ) * ⟪dx, A x⟫ by {
+    simp[loss];
+    fun_trans; -- 'trans' makes me think of 'transitive'.
+    simp;
+    funext dx;
+    -- ⊢ ⟪dx, A x⟫ + ⟪x, A dx⟫ = 2 * ⟪dx, A x⟫
+    -- how do I prove this? The natural proof for me is to do this componentwise
```
Comment (bollu) [inline examples/CircleOptimisation.lean:43]:
```
⊢ ⟪dx, A x⟫ + ⟪x, A dx⟫ = 2 * ⟪dx, A x⟫
```

How do I prove this within SciLean? The natural proof for me is to do this componentwise

File: examples/CircleOptimisation.lean [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+def learning_rate : ℝ := 0.01
+
+noncomputable def TrainLog.nextPoint (step : TrainLog) : ℝ^{2} :=
+  circle_project <| step.point - learning_rate * step.gradient_circle
+
+/-
+TODO: how do I now compute with the `noncomputable` grad_loss?
+-/
+noncomputable def plot_loss (init: ℝ^{2}) (nsteps : ℕ) :
```
Comment (bollu) [inline examples/CircleOptimisation.lean:96]: How do I now get a computable version of this that I can execute?

File: examples/CircleOptimisation.lean [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+/-- choose a particular A -/
+-- the function (x, y) -> (x + y, x - y)
+def A (x : ℝ^{2}) : ℝ^{2} :=
+  ⊞ i, if i = 0 then x[0] + x[1] else x[0] - x[1]
+
+-- set_option trace.Meta.Tactic.fun_trans.rewrite true in
+function_properties A (x : ℝ^{2})
+argument x
+  IsLin, HasAdjoint, IsSmooth, HasAdjDiff,
```
Comment (bollu) [inline examples/CircleOptimisation.lean:23]: Where is `HasAdjoint` used by `SciLean`?

Comment (lecopivo) [reply examples/CircleOptimisation.lean:23]: All transformation rules for `†` are conditioned by `HasAdjoint`.

Similarly `∂` is conditioned by `IsSmooth`
and `∂†` is conditioned by `HasAdjDiff`

Because I'm working with generalization of Hilbert spaces, linearity of a map does not imply existence of adjoint. 
The canonical example is  constant function over natural numbers:
```
(λ (x : ℝ) (y : Nat) => x) : ℝ → (Nat → ℝ)
```
Morally adjoint is sum
```
sum : (Nat → ℝ) → ℝ 
```
but this is not well defined function as not every sequence is summable.

Why do I need this? I really want `(d/dx)† = -d/dx` where
```
d/dx : C∞(ℝ,ℝ) → C∞(ℝ,ℝ)
```
but `C∞(ℝ,ℝ)` is not Hilbert space

File: examples/CircleOptimisation.lean [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+set_option trace.Meta.Tactic.fun_trans.rewrite true in
+def loss (x : ℝ^{2}) : ℝ := ⟪x, A x⟫
+
+function_properties loss (x : ℝ^{2})
+argument x
+  IsSmooth, HasAdjDiff,
+  abbrev ∂ := λ dx => (2 : ℝ) * ⟪dx, A x⟫ by {
+    simp[loss];
+    fun_trans; -- 'trans' makes me think of 'transitive'.
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:39]: Yeah maybe I rename it to `function_transformation` and then have specialized versions called `autodiff` or `symdiff` with slightly different configurations.

File: examples/CircleOptimisation.lean [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    simp;
+    funext dx;
+    -- ⊢ ⟪dx, A x⟫ + ⟪x, A dx⟫ = 2 * ⟪dx, A x⟫
+    -- how do I prove this? The natural proof for me is to do this componentwise
+    sorry
+  }
+
+noncomputable def grad_loss (x : ℝ^{2}) : ℝ^{2} := ∇ loss x
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:47]: You can add a property to `loss`
```
  abbrev ∂† by unfold gradient; unfold loss; fun_trans; simp
```

Then you can define `grad_loss` as
```
def grad_loss (x : ℝ^{2}) : ℝ^{2} := (∇ loss x) rewrite_by unfold gradient; unfold loss; fun_trans
```
and it is computable now.

File: examples/CircleOptimisation.lean [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  gradient_ambient : ℝ^{2} -- gradient at current point
+  gradient_circle : ℝ^{2} -- gradient at current point
+
+
+instance : HDiv (ℝ ^{n}) ℝ (ℝ^{n}) where
+  hDiv v r := ⊞ i, v[i] / r
+
+instance : HMul ℝ (ℝ ^{n}) (ℝ^{n}) where
+  hMul r v := ⊞ i, v[i] * r
+
```
Comment (lecopivo) [inline examples/CircleOptimisation.lean:62]: This is already available with `r • v`

File: examples/CircleOptimisation.lean (hunk 10 above)
Comment (bollu) [reply examples/CircleOptimisation.lean:54]: Thanks! I am unsure why there are three things I can define: `∂`, `†`, and ` ∂†`. My impression is that if all I want is reverse mode AD, I can get by with just ` ∂†`?

Comment (lecopivo) [reply examples/CircleOptimisation.lean:54]: Yes that is correct but for proper reverse mode AD you actually want ℛ
```
ℛ f = (f x, λ dy => ∂† f x dy)
```

**Comment by bollu** [comment]
Merged manually at https://github.com/lecopivo/SciLean/commit/efe0a7a52f672d769a82e80c3b8ecb8d61bac60f
